### PR TITLE
Feat: Implement APIs for Tracking Departed Users with Assigned Tasks

### PR DESCRIPTION
**Pull request by VinuB-Dev**
Date: 15/11/2024
Developer Name: Vignesh B S

---

## Issue Ticket Number
Closes Real-Dev-Squad/website-dashboard#888

## Description

Implemented changes in users API for Tracking Departed Users.

- `users?departed=true` - Retrieves users who had a task assigned and have departed the Discord server.

Added test cases for the above api.

### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [x] Yes
- [ ] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots
<details>
<summary>Screenshot 1</summary>

Screenshot 2024-11-15 at 3 34 37 PM

</details>

## Test Coverage
<details>
<summary>Screenshot 1</summary>

Screenshot 2024-11-15 at 3 33 11 PM

</details>

## Additional Notes

[PRD](https://docs.google.com/document/d/1NCjJQ6RJlQGqdZyr8r2myuaWekYIoEORLr94fKRaCvo/edit?addon_store&tab=t.0#heading=h.cggvkhhvicot)
[Design Doc](https://docs.google.com/document/d/1ZYhy21DLHtTnXfKRlDH-7EVlKrMV2PyZs1_6czTozDk/edit?addon_store&tab=t.0#heading=h.cggvkhhvicot)
Have moved the archieveUsers function to models as it was directly interacts with the database and performing operations such as updating task documents. This was also causing circular dependency issues.

### CODE REVIEW

**Comment by pankajjs** [comment]
why not under feature flag?

File: test/unit/models/tasks.test.js [hunk 1]
```diff
@@ -332,8 +332,8 @@ describe("tasks", function () {
 
     it("Should update task status COMPLETED to DONE", async function () {
       const res = await tasks.updateTaskStatus();
-      expect(res.totalTasks).to.be.equal(8);
-      expect(res.totalUpdatedStatus).to.be.equal(8);
+      expect(res.totalTasks).to.be.equal(12);
+      expect(res.totalUpdatedStatus).to.be.equal(12);
```
Comment (pankajjs) [inline test/unit/models/tasks.test.js:336]: why this change?

File: test/unit/models/tasks.test.js [hunk 2]
```diff
@@ -303,12 +303,12 @@ describe("tasks", function () {
       overdueTask.endsOn = Date.now() / 1000 + 24 * 60 * 60 * 7;
       await tasks.updateTask(overdueTask);
       const usersWithOverdueTasks = await tasks.getOverdueTasks(days);
-      expect(usersWithOverdueTasks.length).to.be.equal(5);
+      expect(usersWithOverdueTasks.length).to.be.equal(8);
```
Comment (pankajjs) [inline test/unit/models/tasks.test.js:306]: why this change?

File: test/unit/models/tasks.test.js [hunk 3]
```diff
@@ -303,12 +303,12 @@ describe("tasks", function () {
...
     });
 
     it("should return all users which have overdue tasks if days is not passed", async function () {
       const usersWithOverdueTasks = await tasks.getOverdueTasks();
-      expect(usersWithOverdueTasks.length).to.be.equal(4);
+      expect(usersWithOverdueTasks.length).to.be.equal(7);
```
Comment (pankajjs) [inline test/unit/models/tasks.test.js:311]: why this change?

File: test/unit/services/tasks.test.js [hunk 4]
```diff
@@ -46,7 +46,7 @@ describe("Tasks services", function () {
       const res = await updateTaskStatusToDone(tasks);
 
       expect(res).to.deep.equal({
-        totalUpdatedStatus: 8,
+        totalUpdatedStatus: 12,
```
Comment (pankajjs) [inline test/unit/services/tasks.test.js:49]: same

File: test/unit/services/users.test.js [hunk 5]
```diff
@@ -76,7 +76,7 @@ describe("Users services", function () {
       expect(res).to.deep.equal({
         message: "Firebase batch operation failed",
         totalUsersArchived: 0,
-        totalOperationsFailed: 20,
+        totalOperationsFailed: 23,
```
Comment (pankajjs) [inline test/unit/services/users.test.js:79]: same

File: test/unit/services/users.test.js [hunk 6]
```diff
@@ -55,7 +55,7 @@ describe("Users services", function () {
 
       expect(res).to.deep.equal({
         message: "Successfully completed batch updates",
-        totalUsersArchived: 20,
```
Comment (pankajjs) [inline test/unit/services/users.test.js:58]: same

File: test/unit/services/tasks.test.js [hunk 7]
```diff
@@ -66,7 +66,7 @@ describe("Tasks services", function () {
 
       expect(res).to.deep.equal({
         totalUpdatedStatus: 0,
-        totalOperationsFailed: 8,
+        totalOperationsFailed: 12,
```
Comment (pankajjs) [inline test/unit/services/tasks.test.js:69]: same

File: controllers/users.js [hunk 8]
```diff
@@ -1029,6 +1029,16 @@ const updateUsernames = async (req, res) => {
   }
 };
 
+const getUsersWithAbandonedTasks = async (req, res) => {
+  try {
+    const data = await userQuery.fetchUsersWithAbandonedTasks();
```
Comment (pankajjs) [inline controllers/users.js:1034]: why are we interacting with model directly? we have data service layer.

File: test/unit/models/tasks.test.js (hunk 1 above)
Comment (VinuB-Dev) [reply test/unit/models/tasks.test.js:336]: I increased the tasks in the mock data to include the required data for the new API. This addition increased the task count by 4. While verifying the test cases, I noticed some were failing due to the changes in mock data, so I updated and fixed them accordingly.

File: test/unit/models/tasks.test.js (hunk 3 above)
Comment (VinuB-Dev) [reply test/unit/models/tasks.test.js:311]: I increased the users in the mock data to include the required data for the new API. This addition increased the user count by 3. While verifying the test cases, I noticed some were failing due to the changes in mock data, so I updated and fixed them accordingly.

Comment (pankajjs) [reply test/unit/models/tasks.test.js:311]: Why are we modifying the tests which are already working? Can't we create another variable for users mock data?

File: controllers/users.js (hunk 8 above)
Comment (VinuB-Dev) [reply controllers/users.js:1034]: I followed the convention currently used in the controller files, where database interaction is handled via models. The model function is then used in the controller and subsequently in routes.
Should I move this part to services instead? I noticed that other functions in the controllers are also following the same approach mentioned above

Comment (pankajjs) [reply controllers/users.js:1034]: we have both service and data-access layer.
It would be good if we separate the concerns.

File: test/unit/models/tasks.test.js (hunk 3 above)
Comment (VinuB-Dev) [reply test/unit/models/tasks.test.js:311]: I can create a separate mock data. My initial approach of modifying the existing fixtures was based on the pattern I observed in our codebase, where previous additions were made to the same fixtures file to keep mock data centralised and easier to maintain.

File: controllers/users.js (hunk 8 above)
Comment (VinuB-Dev) [reply controllers/users.js:1034]: Are you suggesting to move the code in models/user to services/user?
The highlighted code is a controller which used the function in model to get the result.

Comment (pankajjs) [reply controllers/users.js:1034]: I am suggesting to follow controller, service, data-access layer pattern.
Data-access layer should use the model. Service should use data-access layer and so on.

Comment (VinuB-Dev) [reply controllers/users.js:1034]: Okay got it. Will make this change in both PR's.

**Comment by vinit717** [comment]
Are pr 2251 and 2252 stack PR? and In what order I need to review them?

**Comment by VinuB-Dev** [comment]
> Are pr 2251 and 2252 stack PR? and In what order I need to review them?

There were 2 api's in the original PR, I have separated both of them into separate PR's. 
Each one can be reviewed independently.

File: test/unit/models/tasks.test.js (hunk 3 above)
Comment (VinuB-Dev) [reply test/unit/models/tasks.test.js]: Have added the changes @pankajjs

File: controllers/users.js (hunk 8 above)
Comment (VinuB-Dev) [reply controllers/users.js]: Have added the changes @pankajjs

File: models/tasks.js [hunk 9]
```diff
@@ -701,6 +701,20 @@ const markUnDoneTasksOfArchivedUsersBacklog = async (users) => {
   }
 };
 
+const fetchIncompleteTaskForUser = async (user) => {
+  const COMPLETED_STATUSES = [DONE, COMPLETED];
+  try {
+    const incompleteTaskForUser = await tasksModel
```
Comment (pankajjs) [inline models/tasks.js:707]: Can't we just pass id no need to pass user object right?

File: models/users.js [hunk 10]
```diff
@@ -27,6 +31,46 @@ const { formatUsername } = require("../utils/username");
 const { logType } = require("../constants/logs");
 const { addLog } = require("../services/logService");
 
+const archiveUsers = async (usersData) => {
```
Comment (pankajjs) [inline models/users.js:34]: Write doc for this function. Take reference from addOrUpdate.

File: models/users.js [hunk 11]
```diff
@@ -1030,7 +1074,21 @@ const updateUsersWithNewUsernames = async () => {
   }
 };
 
+const fetchUsersNotInDiscordServer = async () => {
```
Comment (pankajjs) [inline models/users.js:1077]: Write doc for this too.

File: routes/users.js [hunk 12]
```diff
@@ -67,6 +68,7 @@ router.patch("/profileURL", authenticate, userValidator.updateProfileURL, users.
 router.patch("/rejectDiff", authenticate, authorizeRoles([SUPERUSER]), users.rejectProfileDiff);
 router.patch("/:userId", authenticate, authorizeRoles([SUPERUSER]), users.updateUser);
 router.get("/suggestedUsers/:skillId", authenticate, authorizeRoles([SUPERUSER]), users.getSuggestedUsers);
+
```
Comment (pankajjs) [inline routes/users.js:71]: remove this

File: models/tasks.js [hunk 13]
```diff
@@ -701,6 +701,20 @@ const markUnDoneTasksOfArchivedUsersBacklog = async (users) => {
   }
 };
 
+const fetchIncompleteTaskForUser = async (user) => {
```
Comment (pankajjs) [inline models/tasks.js:704]: Write doc for this too.

File: test/unit/models/users.test.js [hunk 14]
```diff
@@ -527,4 +528,24 @@ describe("users", function () {
       expect(userDoc.user.roles.super_user).to.be.equal(false);
     });
   });
+
+  describe("fetchUsersNotInDiscordServer", function () {
+    beforeEach(async function () {
+      // Clean the database
```
Comment (pankajjs) [inline test/unit/models/users.test.js:534]: remove comment

File: test/unit/models/users.test.js [hunk 15]
```diff
@@ -527,4 +528,24 @@ describe("users", function () {
...
+      await cleanDb();
+
+      // Add test users to the database
```
Comment (pankajjs) [inline test/unit/models/users.test.js:537]: same

File: test/unit/models/users.test.js [hunk 16]
```diff
@@ -527,4 +528,24 @@ describe("users", function () {
       expect(userDoc.user.roles.super_user).to.be.equal(false);
     });
   });
+
+  describe("fetchUsersNotInDiscordServer", function () {
```
Comment (pankajjs) [inline test/unit/models/users.test.js:532]: what about the negative test?

**Review by pankajjs — COMMENTED** [review]
Left few comments @VinuB-Dev

File: models/tasks.js (hunk 9 above)
Comment (VinuB-Dev) [reply models/tasks.js:707]: Yes, I'll make the changes.

File: test/unit/models/users.test.js (hunk 16 above)
Comment (VinuB-Dev) [reply test/unit/models/users.test.js:532]: Have added it

File: models/users.js (hunk 11 above)
Comment (VinuB-Dev) [reply models/users.js:1088]: Done

File: models/users.js (hunk 10 above)
Comment (VinuB-Dev) [reply models/users.js:40]: Done

File: test/unit/services/users.test.js [hunk 17]
```diff
@@ -100,4 +105,54 @@ describe("Users services", function () {
       expect(newUsername).to.equal("john-doe-1");
     });
   });
+
+  describe("fetchUsersWithAbandonedTasks", function () {
```
[the remark this answers is not in the document]
Comment (VinuB-Dev) [reply test/unit/services/users.test.js:109]: Done

File: models/users.js [hunk 18]
```diff
@@ -1030,7 +1080,26 @@ const updateUsersWithNewUsernames = async () => {
   }
 };
 
+/**
+ * Fetches users who are not in the Discord server.
+ * @returns {Promise<FirebaseFirestore.QuerySnapshot>} - A promise that resolves to a Firestore QuerySnapshot containing the users matching the criteria.
```
Comment (pankajjs) [inline models/users.js:1085]: Simplify the comment

File: services/users.js [hunk 19]
```diff
@@ -1,44 +1,29 @@
...
-    summary.failedUserDetails = [...usersBatch];
-    return { message: USERS_PATCH_HANDLER_ERROR_MESSAGES.ARCHIVE_USERS.BATCH_DATA_UPDATED_FAILED, ...summary };
+    const eligibleUsersWithTasks = [];
+
+    const userSnapshot = await fetchUsersNotInDiscordServer();
+
+    for (const userDoc of userSnapshot.docs) {
+      const user = userDoc.data();
+
+      // Check if the user has any tasks with status not in [Done, Complete]
```
Comment (pankajjs) [inline services/users.js:16]: remove comment

File: test/unit/services/users.test.js [hunk 20]
```diff
@@ -100,4 +105,49 @@ describe("Users services", function () {
...
+
+      const activeUser = abandonedUsersData[2];
+      await userModel.add(activeUser);
+
+      const activeTask = abandonedTasksData[3];
+      await tasksModel.add(activeTask);
+      const usersWithAbandonedTasks = await getUsersWithIncompleteTasks();
+      expect(usersWithAbandonedTasks).to.be.an("array");
+      expect(usersWithAbandonedTasks).to.have.lengthOf(0);
+    });
```
Comment (pankajjs) [inline test/unit/services/users.test.js:151]: add more asserts

File: test/unit/models/users.test.js [hunk 21]
```diff
@@ -527,4 +528,43 @@ describe("users", function () {
...
+      sinon.restore();
+    });
+
+    it("should fetch users not in discord server", async function () {
+      const usersNotInDiscordServer = await users.fetchUsersNotInDiscordServer();
+      expect(usersNotInDiscordServer.docs.length).to.be.equal(2);
+    });
+
+    it("should return an empty array if there are no users in the database", async function () {
+      // Using the active user from our test data
```
Comment (pankajjs) [inline test/unit/models/users.test.js:551]: remove comment

**Comment by pankajjs** [comment]
Left few comments @VinuB-Dev

**Comment by pankajjs** [comment]
Test missing for controller

File: test/unit/services/users.test.js (hunk 20 above)
Comment (VinuB-Dev) [reply test/unit/services/users.test.js:151]: The test case checks whether an empty array is returned when there are no users with abandoned tasks, and the existing assertions validate this behaviour. Could you clarify what kind of additional assertions you are expecting?

**Comment by VinuB-Dev** [comment]
> Test missing for controller

I can't find a controller test case file under test folder, where should I add the test case for controller?

**Comment by pankajjs** [comment]
> > Test missing for controller
> 
> I can't find a controller test case file under test folder, where should I add the test case for controller?

Integration

File: test/unit/services/users.test.js (hunk 20 above)
Comment (pankajjs) [reply test/unit/services/users.test.js:151]: Resolved

File: services/users.js (hunk 19 above)
Comment (VinuB-Dev) [reply services/users.js]: Done

File: test/unit/models/users.test.js (hunk 21 above)
Comment (VinuB-Dev) [reply test/unit/models/users.test.js]: Done

File: test/integration/users.test.js [hunk 22]
```diff
@@ -2646,4 +2652,42 @@ describe("Users", function () {
       expect(res).to.have.status(401);
     });
   });
+
+  describe("GET /users/departed-users", function () {
```
Comment (pankajjs) [inline test/integration/users.test.js:2656]: test missing for error response? @VinuB-Dev

**Comment by pankajjs** [comment]
Not under feature flag? @VinuB-Dev i

**Comment by VinuB-Dev** [comment]
@vinit717  Should we put these APIs under a feature flag?

The APIs are read-only, fetching data from the database without performing any write operations. They are completely independent, involve straightforward database queries, and are not dependent on external factors. Considering their simplicity, I feel we don’t need to place them under a feature flag.

File: test/integration/users.test.js (hunk 22 above)
Comment (VinuB-Dev) [reply test/integration/users.test.js:2656]: As discussed, I’ve tried stubbing both the service and model layers, but the failure state doesn’t get triggered as expected. Have added a message in discord to check if there is any other way to achieve this.
Will update here if I get a solution.

**Comment by pankajjs** [comment]
@VinuB-Dev You have moved archivedUser function to different file. Have you fixed all the imports statement?

**Comment by VinuB-Dev** [comment]
> @VinuB-Dev You have moved archivedUser function to different file. Have you fixed all the imports statement?

Yes I have and all test cases are working fine.

File: controllers/users.js [hunk 23]
```diff
@@ -1029,6 +1030,17 @@ const updateUsernames = async (req, res) => {
   }
 };
 
+const getUsersWithAbandonedTasks = async (req, res) => {
+  try {
+    const data = await userService.getUsersWithIncompleteTasks();
+    if (data.length === 0) res.status(204).send();
```
Comment (pankajjs) [inline controllers/users.js:1036]: why this?

Comment (VinuB-Dev) [reply controllers/users.js:1036]: This is the case which we were discussing yesterday in hangout, if the response is empty then send 204.

Comment (pankajjs) [reply controllers/users.js:1036]: The response is not empty. It has a list which is empty right?

Comment (VinuB-Dev) [reply controllers/users.js:1036]: Yes, the response previous was
```
{
  "message": "No users with abandoned tasks found",
  "data": []
}
```
Since there is no data to send to the user, I have replaced it with 204 based on yesterday's discussions.

**Review by pankajjs — APPROVED** [review]

**Comment by tejaskh3** [comment]
as this PR is outdated compared to develop, Could you please pull the latest changes from develop into your branch and resolve any conflicts if they arise?

**Comment by VinuB-Dev** [comment]
> as this PR is outdated compared to develop, Could you please pull the latest changes from develop into your branch and resolve any conflicts if they arise?

Done

File: controllers/users.js [hunk 24]
```diff
@@ -192,6 +192,32 @@ const getUsers = async (req, res) => {
...
+              prev: result.prevId ? getPaginationLink(req.query, "prev", result.prevId) : "",
+            },
+          });
+        } catch (error) {
+          logger.error("Error when fetching users who abandoned tasks:", error);
+          return res.boom.badImplementation(INTERNAL_SERVER_ERROR);
+        }
+      } else {
+        return res.boom.notFound("Route not found");
+      }
```
Comment (pankajjs) [inline controllers/users.js:218]: use early return

File: controllers/users.js [hunk 25]
```diff
@@ -192,6 +192,32 @@ const getUsers = async (req, res) => {
...
 
+    const isDeparted = req.query.departed === "true";
+
+    if (isDeparted) {
+      let result;
+      if (dev) {
+        try {
+          result = await dataAccess.retrieveUsers({ query: req.query });
+          const departedUsers = await userService.getUsersWithIncompleteTasks(result.users);
+          if (departedUsers.length === 0) res.status(204).send();
```
Comment (pankajjs) [inline controllers/users.js:203]: `
if (departedUsers.length === 0){
return res.status(204).send();
}
`
Can be written like this.

**Review by pankajjs — CHANGES_REQUESTED** [review]
Left few comments

File: controllers/users.js [hunk 26]
```diff
@@ -191,6 +192,32 @@ const getUsers = async (req, res) => {
       }
     }
 
+    const isDeparted = req.query.departed === "true";
+
+    if (isDeparted) {
+      let result;
+      if (dev) {
+        try {
+          result = await dataAccess.retrieveUsers({ query: req.query });
```
Comment (pankajjs) [inline controllers/users.js:201]: what will be the query here?

Comment (VinuB-Dev) [reply controllers/users.js:201]: It'll be `{ departed: "true" }`

File: controllers/users.js [hunk 27]
```diff
@@ -192,6 +192,32 @@ const getUsers = async (req, res) => {
...
+            message: "Users with abandoned tasks fetched successfully",
+            users: departedUsers,
+            links: {
+              next: result.nextId ? getPaginationLink(req.query, "next", result.nextId) : "",
+              prev: result.prevId ? getPaginationLink(req.query, "prev", result.prevId) : "",
+            },
+          });
+        } catch (error) {
+          logger.error("Error when fetching users who abandoned tasks:", error);
+          return res.boom.badImplementation(INTERNAL_SERVER_ERROR);
```
[the remark this answers is not in the document]
Comment (VinuB-Dev) [reply controllers/users.js:214]: It's be 500
You can see ti at the end here - https://www.npmjs.com/package/boom/v/2.5.0?activeTab=readme

File: controllers/users.js [hunk 28]
```diff
@@ -191,6 +192,30 @@ const getUsers = async (req, res) => {
...
     }
 
+    const isDeparted = req.query.departed === "true";
+
+    if (isDeparted) {
+      if (!dev) {
+        return res.boom.notFound("Route not found");
+      }
+      try {
+        const result = await dataAccess.retrieveUsers({ query: req.query });
```
Comment (pankajjs) [inline controllers/users.js:202]: What is it returning?

**Review by Achintya-Chatterjee — COMMENTED** [review]
please write API contract for this API

File: controllers/users.js (hunk 28 above)
Comment (VinuB-Dev) [reply controllers/users.js:202]: An object
`{ users: [], nextId: '' , prevId: ''}`

**Comment by VinuB-Dev** [comment]
> please write API contract for this API

@Achintya-Chatterjee Where should I add it? In the description or somewhere else?

**Comment by Achintya-Chatterjee** [comment]
> > please write API contract for this API
> 
> @Achintya-Chatterjee Where should I add it? In the description or somewhere else?

[https://github.com/Real-Dev-Squad/website-api-contracts](you can add it here)

**Comment by VinuB-Dev** [comment]
Splitting this PR into smaller parts for easier review. This PR will serve as the main split: https://github.com/Real-Dev-Squad/website-backend/pull/2267. Closing this PR as part of the process.
